playbooks/adhoc/upgrades/library/openshift_upgrade_config.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# vim: expandtab:tabstop=4:shiftwidth=4
+
+import os
+import shutil
+import yaml
+
+from datetime import datetime
+
+"""Ansible module for modifying OpenShift configs during an upgrade"""
+
+DOCUMENTATION = '''
+---
+module: openshift_upgrade_config
+short_description: OpenShift Upgrade Config
+author: Jason DeTiberus
+requirements: [ ]
+'''
+EXAMPLES = '''
+'''
+
+def get_cfg_dir():
+    cfg_path = '/etc/origin/'
+    if not os.path.exists(cfg_path):
+        cfg_path = '/etc/openshift/'
+    return cfg_path
+
+
+def upgrade_master_3_0_to_3_1(backup):
+    changed = False
+
+    # Facts do not get transferred to the hosts where custom modules run,
+    # need to make some assumptions here.
+    master_config = os.path.join(get_cfg_dir(), 'master/master-config.yaml')
+
+    f = open(master_config, 'r')
+    config = yaml.safe_load(f.read())
+    f.close()
+
+    # Remove v1beta3 from apiLevels:
+    if 'apiLevels' in config and \
+        'v1beta3' in config['apiLevels']:
+            config['apiLevels'].remove('v1beta3')
+            changed = True
+    if 'apiLevels' in config['kubernetesMasterConfig'] and \
+        'v1beta3' in config['kubernetesMasterConfig']['apiLevels']:
+            config['kubernetesMasterConfig']['apiLevels'].remove('v1beta3')
+            changed = True
+
+    # Add the new master proxy client certs:
+    if 'proxyClientInfo' not in config['kubernetesMasterConfig']:
+        config['kubernetesMasterConfig']['proxyClientInfo'] = {
+            'certFile': 'master.proxy-client.crt',
+            'keyFile': 'master.proxy-client.key'
+        }
+
+    if changed:
+        if backup:
+            timestamp = datetime.now().strftime('%Y%m%d%H%M%S')
+            basedir = os.path.split(master_config)[0]
+            backup_file = os.path.join(basedir, 'master-config.yaml.bak-%s'
+                % timestamp)
+            shutil.copyfile(master_config, backup_file)
```

Comment:
ansible module_utils provides a backup_local method to handle time-stamped backups, I would recommend using that instead. https://github.com/ansible/ansible/blob/stable-1.9/lib/ansible/module_utils/basic.py#L1311-L1321